File: server/genlib/src/main/java/cn/teclub/ha3/request/StWsMessage.java
package cn.teclub.ha3.request;

import cn.teclub.ha3.utils.StObject;


//@SuppressWarnings("ALL")
@SuppressWarnings("unused")
public class StWsMessage extends StObject
{
    public static final long ID_INVALID = -1;
    public static final long ID_SERVER  = 0;

    public enum Command {
        NULL(0x00),
        APPLY_ADD_DEVICE(0x30),
        GRANT_ADD_DEVICE(0x31),
        ECHO(0x40),
        UPDATE_FRIEND(0x41),
        MAKE_CALL(0x42),
        ACCEPT_CALL(0x43),
        HANDUP_CALL(0x44),
        PING(0x45),
        GET_CALL_INFO(0x46),
        GET_DEV_INFO(0x47),
        RECORD_CAPTURE(0X48),
        REQ_MAX(0x50),

        CALL_OFFER(0x51),
        CALL_ANSWER(0x52),
        CALL_CANDIDATE(0x53),
        CALL_BYE(0x54),
        CALL_CANDIDATE_RM(0x55),
        SEND_DATA(0x60),



        @Deprecated
        VIDEO_CALL(0x0A01),
        MAX(0xFFFF);

        private short index;

        Command(int idx) {
            this.index = (short) idx;
        }

        public short getIndex() {
            return index;
        }
    }

    // [Theodor: 2019-04-10] currently not used
    public enum Type{
        None, Request, Allow, Deny
    }

    private short   cmd;
    private Type    type;
    private long    ssid;

    private long    src, dst;
    private int   len;
    private String  info;
    private String  opt; // for debug & test


    /*
    public static StWsMessage buildRequest( final Command cmd, final String info, final long from) {
        // todo: check cmd is valid
        return new StWsMessage(cmd, Type.Request, info, from, ID_SERVER);
    }


    public static StWsMessage buildRequest( final Command cmd, final String info, final long from, final long to) {
        return new StWsMessage(cmd, Type.Request, info, from, to );
    }*/


    public StWsMessage buildResponse(final Type t, final String info){
        return new StWsMessage(this.ssid, getCommand(), t, info, dst, src );
    }


    // -----------------------------------------------------------------------------
    // constructors
    // -----------------------------------------------------------------------------

    /**
     * default constructor: required by json
     */
    public StWsMessage(){
    }


    /**
     * constructor: called by application
     */
    public StWsMessage(final long ssid, final Command cmd, final Type t, final String info, final long from, final long to) {
        this.ssid = ssid;
        this.cmd = cmd.getIndex();
        this.type = t;
        this.src = from;
        this.dst = to;
        if(info == null) {
            this.info = "";

        }else{
            this.len = info.length();
            this.info = info;
        }
        this.opt = "";
        this.opt += "TODO";
    }


    public StWsMessage(final long ssid, final String info, final long from, final long to) {
        this(ssid, Command.NULL, Type.None, info, from, to);
    }


    public StWsMessage(final long ssid, final Command cmd, final Type t, final String info){
        this(ssid, cmd, t, info, 0x00, 0x01);
    }



    // -----------------------------------------------------------------------------
    // methods & properties
    // -----------------------------------------------------------------------------

    // for debug & test
    public void addNote(String note){
        this.opt += note;
    }


    public String toString(){
        return "[" +  getClass().getSimpleName() +  "]" + getCommand() + "("+
                "0x" + util.to4CharHex(cmd) + ")," + type + "," +
                ",ssid:" + util.to16CharHex(ssid) +
                "," + util.to16CharHex(src) + "->" + util.to16CharHex(dst) +
                "," + len + "B," + info + "[opt]" + opt;

    }


    public long getSsid() {
        return ssid;
    }

    public void setSsid(long ssid) {
        this.ssid = ssid;
    }


    public short getCmd() {
        return cmd;
    }


    public Command getCommand(){
        Command[] list = Command.values();
        for(Command c : list) {
            if(c.getIndex() == this.cmd) {
                return c;
            }
        }
        return Command.NULL;
    }


    public void setCmd(short cmd) {
        this.cmd = cmd;
    }

        public Type getType() {
        return type;
    }

    public void setType(Type type) {
        this.type = type;
    }


    public int getLen() {
        return len;
    }

    public void setLen(int len) {
        this.len = len;
    }

    public String getInfo() {
        return info;
    }

    public void setInfo(String info) {
        this.info = info;
    }

    public long getSrc() {
        return src;
    }

    public void setSrc(long src) {
        this.src = src;
    }

    public long getDst() {
        return dst;
    }

    public void setDst(long dst) {
        this.dst = dst;
    }


    public String getOpt() {
        return opt;
    }

    public void setOpt(String opt) {
        this.opt = opt;
    }
}

File: test_for_fs/test_for_fs/StWsMessage.swift
//
//  StWsMessage.swift
//  test_for_fs
//
//  Created by qians on 2019/12/11.
//  Copyright © 2019 qians. All rights reserved.
//
import Foundation

class StWsMessage: NSObject{
    
    static let ID_INVALID = -1;
    static let ID_SERVER  = 0;
    
    enum Command: Int{
        
        case NULL = 0x00
        case APPLY_ADD_DEVICE = 0x30
        case GRANT_ADD_DEVICE = 0x31
        case ECHO = 0x40
        case UPDATE_FRIEND = 0x41
        case MAKE_CALL = 0x42
        case ACCEPT_CALL = 0x43
        case HANDUP_CALL = 0x44
        case PING = 0x45
        case GET_CALL_INFO = 0x46
        case GET_DEV_INFO = 0x47
        case RECORD_CAPTURE = 0x48
        case REQ_MAX = 0x50

        case CALL_OFFER = 0x51
        case CALL_ANSWER = 0x52
        case CALL_CANDIDATE = 0x53
        case CALL_BYE = 0x54
        case CALL_CANDIDATE_RM = 0x55
        case SEND_DATA = 0x60

        case MAX = 0xFFFF
    }
    enum stType: String{
        case None = "None"
        case Request = "Request"
        case Allow = "Allow"
        case Deny = "Deny"
    }
    
    var ssid: Int
    var src: String
    var dst: String
    var len: Int
    var info: String
    
    // for debug & test
    var opt: String
    
    var cmd: Command
    var type: stType
    
//    init(AnyObject: msg){
//        
//    }
    
    init(ssid: Int, cmd: Command, type: stType, info: String, from: String, to: String) {
        self.ssid = ssid
        self.cmd = cmd
        self.type = type
        self.info = info
        self.src = from
        self.dst = to
        
        if info.isEmpty{
            self.info = ""
            self.len = 0
        }else{
            self.len = info.lengthOfBytes(using: String.Encoding.utf8)
        }
        self.opt = ""
        self.opt += "TODO"
    }

    convenience init(ssid: Int, info: String, from: String, to: String) {
        self.init(ssid: ssid, cmd: Command.NULL, type: stType.None, info: info, from: from, to: to)
    }
    
    convenience init(ssid: Int, cmd: Command, type: stType, info: String) {
        self.init(ssid: ssid, cmd: cmd, type: type, info: info, from: "0", to: "1")
    }
    
    func addNote(note: String) -> Void{
        self.opt += note
    }
    
    func toString() -> String {
        var dict: [String: Any] = [:]
        
        dict["ssid"] = ssid
        dict["cmd"]  = cmd.rawValue
        dict["type"] = type.rawValue
        dict["info"] = info
        dict["len"]  = len
        dict["opt"]  = opt
        dict["src"]  = src
        dict["dst"]  = dst
        do {
            let theJSONData = try JSONSerialization.data(withJSONObject: dict, options: [])
            let theJSONText = String(data: theJSONData, encoding: String.Encoding.utf8)
            
            return theJSONText!
        }catch {
            print("error serializing JSON: \(error)")
            return ""
        }
       
    }
    
    func getSsid() -> Int {
        return self.ssid
    }
    
    func setSsid(ssid: Int) -> Void {
        self.ssid = ssid
    }
    
    func getCmd() -> Command {
        return self.cmd
    }
    
    func setCmd(cmd: Command) -> Void {
        self.cmd = cmd
    }
    
    func getType() -> stType {
        return self.type
    }
    
    func setType(type: stType) -> Void {
        self.type = type
    }
    
    func getLen() -> Int {
        return self.len
    }
    
    func setLen(len: Int) -> Void {
        self.len = len
    }
    
    func getInfo() -> String {
        return self.info
    }
    
    func setInfo(info: String) -> Void {
        self.info = info
    }
    
    func getSrc() -> String {
        return self.src
    }
    
    func setSrc(src: String) -> Void {
        self.src = src
    }
    
    func getDst() -> String {
        return self.dst
    }
    
    func setDst(dst: String) -> Void {
        self.dst = dst
    }
    
    func getOpt() -> String {
        return self.opt
    }
    
    func setOpt(opt: String) -> Void {
        self.opt = opt
    }
}

